plugin/documents.py

```diff
@@ -616,46 +625,65 @@ def render_highlights_on_main_thread() -> None:
 
     # --- textDocument/complete ----------------------------------------------------------------------------------------
 
-    def _on_query_completions_async(self, resolve: ResolveCompletionsFn, location: int) -> None:
-        session = self.session('completionProvider', location)
-        if not session:
-            resolve([], 0)
+    def _on_query_completions_async(self, resolve_completion_list: ResolveCompletionsFn, location: int) -> None:
+        sessions = self.sessions('completionProvider')
+        if not sessions:
+            resolve_completion_list([], 0)
             return
         self.purge_changes_async()
-        can_resolve_completion_items = bool(session.get_capability('completionProvider.resolveProvider'))
-        config_name = session.config.name
-        session.send_request_async(
-            Request.complete(text_document_position_params(self.view, location), self.view),
-            lambda res: self._on_complete_result(res, resolve, can_resolve_completion_items, config_name),
-            lambda res: self._on_complete_error(res, resolve))
-
-    def _on_complete_result(self, response: Optional[Union[dict, List]], resolve: ResolveCompletionsFn,
-                            can_resolve_completion_items: bool, session_name: str) -> None:
-        response_items = []  # type: List[Dict]
-        flags = 0
+        completion_promises = []  # type: List[Promise[ResolvedCompletions]]
+        for session in sessions:
+
+            def completion_request() -> Promise[ResolvedCompletions]:
+                return session.send_request_task(
+                    Request.complete(text_document_position_params(self.view, location), self.view)
+                ).then(lambda response: (response, session.config.name))
+
+            completion_promises.append(completion_request())
+        LspResolveDocsCommand.completions = []
```

Comment:
I'm thinking it would be better to clear the list from `_on_all_settled`. In theory, with slow-running requests, it could happen that some stale completions would sneak into the latest request results.

Reply:
> In theory, with slow-running requests, it could happen that some stale completions would sneak into the latest request results.

To add on this, here is how the code currently behaves.

If a language server has `isIncomplete` completions and we: 
>> type "a"
... request#1 pending, set LspResolveDocsCommand.completions = []
<< get completions from server, set LspResolveDocsCommand.completions = response1
>> type "b"
... request#2 pending, set LspResolveDocsCommand.completions = []
>> user triggers documentation popup for completion item from request#1, before the response for the request "b" was received.

In that case an index error will be thrown. 
In that case moving the clearing logic to `_on_all_settled` would solve the problem.

> I'm thinking it would be better to clear the list from _on_all_settled

But then we would clear the list and then immediately populate it: (why would we clear the list before assigning a new completion, when we can just assign the new completions)
https://github.com/sublimelsp/LSP/pull/1582/files#diff-982c110867d9809c978a5597e4da66ce7538b327675a66b8f27214413b7c0ec1R675

Reply:
Good point. I thought that with incomplete completions the popup will close on typing the next character but it doesn't.

But on the other hand, with the current code, there could be another issue:

With `isIncomplete` completions, the user types "a" and then "b". Both requests are started but are slow to respond so when they both resolve, they both append items to the `LspResolveDocsCommand.completions` list, and triggering documentation could then fail in some cases.

Reply:
There a couple of approaches that came to my mind:

a) never clear the completions. (currently we only clear them when we trigger the AC)
LspResolveDocsCommand.completions = []

b) pass the completion item instead of the index: (feels like the a right approach, but performance might be affected)
```diff
- st_details += make_command_link("lsp_resolve_docs", "More", {"index": index, "session_name": session_name})
+ st_details += make_command_link("lsp_resolve_docs", "More", {"completion_item": completion_item, "session_name": session_name})

```

Reply:
I would choose `a)` because it is simpler and not that different from the current behavoior.

Reply:
> a) never clear the completions.

How can this work if we never clear them?

Reply:
We assign the response to `LspResolveDocsCommand.completions` (keep in mind that I no longer use `LspResolveDocsCommand.completions.entend(response_items)` like I used to. Instead I just use `=` as this will prevent the list from growing to infinity)
```
LspResolveDocsCommand.completions[session_name] = response_items
```
https://github.com/sublimelsp/LSP/pull/1582/commits/01a98131d403ef89c79d442e5b9f36eb28e0c3b7#diff-982c110867d9809c978a5597e4da66ce7538b327675a66b8f27214413b7c0ec1R675

When we want to resolve the completion item, 
we will look up the completion item in the `LspResolveDocsCommand.completions`
```
    def run(self, edit: sublime.Edit, index: int, session_name: str, event: Optional[dict] = None) -> None:
        item = self.completions[session_name][index]
```

https://github.com/sublimelsp/LSP/pull/1582/commits/01a98131d403ef89c79d442e5b9f36eb28e0c3b7#diff-1e31f37d39e534dca2dca7fd81d29d7e5a039a4b55a5765fdb4292c7a87760aaR21

That is it. 

There is no need to clear the `LspResolveDocsCommand.completions = {}` (changed to dict, previously was an array)

We can just delete this line, to fix the index error that would be thrown.
https://github.com/sublimelsp/LSP/pull/1582/commits/01a98131d403ef89c79d442e5b9f36eb28e0c3b7#diff-982c110867d9809c978a5597e4da66ce7538b327675a66b8f27214413b7c0ec1R643

Reply:
Another day, maybe I lost some context already, but clearing from `_on_query_completions_async` seems wrong because of this case:

(isIncomplete == true)
1. A completion request is completed and a completion popup is shown.
2. User types another letter - `LspResolveDocsCommand.completions` is cleared immediately
3. The previous request is slow to respond and the user tries to invoke documentation on some existing item - an error because `LspResolveDocsCommand.completions` is empty.

I think after the change to a dict it makes even more sense to clear it from the `_on_all_settled` because new completion request might not trigger for the same session names so at least from the memory/GC point of view it would make sense to clear that object at some point.

Reply:
I moved it to `_on_all_settled` to avoid errors :)

>  it would make sense to clear that object at some point

I guess the "ideal" place to clear the dict would be when the AC closes(either by the user selecting the completion item, or by pressing the esc key, or by pressing left right arrows). I know that there is a `is_auto_complete_visible()` api, but I don't yet have a simple solution for that approach(so I would avoid complicating the logic now).

But I believe that moving it to `_on_all_settled` improves things a bit.